### PR TITLE
Add an optional exitCode to ExecStatusEvent so clients can use it

### PR DESCRIPTION
**Pull request by dwijnand**

### CODE REVIEW

File: main/src/main/scala/sbt/MainLoop.scala [hunk 1]
```diff
@@ -155,11 +155,28 @@ object MainLoop {
...
+            None
+        }
+    }
     StandardMain.exchange publishEventMessage ExecStatusEvent(
       "Done",
       channelName,
       exec.execId,
-      newState.remainingCommands.toVector map (_.commandLine))
+      newState.remainingCommands.toVector map (_.commandLine),
+      exitCode,
```
Comment (eed3si9n) [inline main/src/main/scala/sbt/MainLoop.scala]: There's also successful event too right?
This needs some rethinking.

Comment (dwijnand) [reply main/src/main/scala/sbt/MainLoop.scala]: you're right.. last second change. will fix tomorrow

**Comment by eed3si9n** [comment]
LSP already has the concept of Response and Error, so I'd say we should handle these in a more idiomatic way.
see also https://github.com/sbt/sbt/issues/3863

**Review by eed3si9n — CHANGES_REQUESTED** [review]
I think we should fold this into #3863

File: main/src/main/scala/sbt/MainLoop.scala [hunk 2]
```diff
@@ -155,13 +155,72 @@ object MainLoop {
...
+      StandardMain.exchange publishEventMessage doneEvent
+    }
     newState
   }
 
   def logFullException(e: Throwable, log: Logger): Unit = State.logFullException(e, log)
-}
+
+  private def determineExitCode(state: State, newState: State): Option[Int] = {
+    sealed trait ExitCode
```
Comment (eed3si9n) [inline main/src/main/scala/sbt/MainLoop.scala]: See https://github.com/sbt/sbt/blob/1.x/protocol/src/main/scala/sbt/internal/langserver/ErrorCodes.scala

**Review by eed3si9n — CHANGES_REQUESTED** [review]
Getting closer.

File: main/src/main/scala/sbt/MainLoop.scala (hunk 2 above)
Comment (dwijnand) [reply main/src/main/scala/sbt/MainLoop.scala]: what are you requesting is changed?

Comment (eed3si9n) [reply main/src/main/scala/sbt/MainLoop.scala]: You're currently making up 0 or 1 based numbering that gets sent to LSP client correct?
http://www.jsonrpc.org/specification determines the number you have to use.

Comment (dwijnand) [reply main/src/main/scala/sbt/MainLoop.scala]: it says:

> The remainder of the space is available for application defined errors.

I'm using 1 when there is some unknown error that was handled by onFailure, that's it.

Comment (eed3si9n) [reply main/src/main/scala/sbt/MainLoop.scala]: VS Code uses -32800 for cancellation. Maybe it's better to extend to some large negative integers following the convention, like `-33000`.

Comment (dwijnand) [reply main/src/main/scala/sbt/MainLoop.scala]: why is `-33000` better than `1` for "unknown error"?

**Comment by dwijnand** [comment]
jfyi, this is green now @eed3si9n

File: main/src/main/scala/sbt/MainLoop.scala (hunk 2 above)
Comment (eed3si9n) [reply main/src/main/scala/sbt/MainLoop.scala]: From the point of view of the LSP client (let's say Vim or something) they need to handle various error events. I want to fit into the ecosystem harmoniously.

**Review by eed3si9n — APPROVED** [review]

**Pull request merged**